src/pybind/mgr/k8sevents/module.py

```diff
@@ -0,0 +1,1072 @@
+    def event_name(self):
+        if self.msg_type == 'heartbeat':
+            return 'mgr.Heartbeat'
+        elif self.healthcheck:
+            return 'mgr.health.{}'.format(self.healthcheck.name)
+        elif self.msg_type == 'audit':
+            return 'mgr.audit.{}'.format(self.cmd).replace(' ', '_')
+        elif self.msg_type == 'config':
+            return 'mgr.ConfigurationChange'
+        elif self.msg_type == 'startup':
+            return "mgr.k8sevents-module"
+        else:
+            return None
+   
+    @property
+    def event_entity(self):
+        if self.msg_type == 'audit':
+            return self.msg.replace("\'","").split('entity=')[1].split(' ')[0]
+        else:
+            return None
+    
+    @property
+    def event_msg(self):
+        if self.msg_type == 'audit':
+            return "Client '{}' issued: ceph {}".format(self.event_entity, self.cmd)
+
+        elif self.healthcheck:
+            return self.healthcheck.text
+        else:
+            return self.msg
+
+class RookCeph(object):
+    """Establish environment defaults when interacting with rook-ceph"""
+
+    pod_name = os.environ['POD_NAME']
+    host = os.environ['NODE_NAME']
+    namespace = os.environ.get('POD_NAMESPACE', 'rook-ceph')
+    cluster_name = os.environ.get('ROOK_CEPH_CLUSTER_CRD_NAME', 'rook-ceph')
+    api = client.CoreV1Api()
+
+class BaseThread(threading.Thread):
+    health = 'OK'
+    reported = False
+    daemon = True
+
+
+class NamespaceWatcher(RookCeph, BaseThread):
+    """Watch events in a given namespace 
+    
+    Using the watch package we can listen to event traffic in the namespace to 
+    get an idea of what kubernetes related events surround the ceph cluster. The
+    thing to bear in mind is that events have a TTL enforced by the kube-apiserver
+    so this stream will only really show activity inside this retention window.
+    """
+
+    def __init__(self, namespace=None):
+        super(NamespaceWatcher, self).__init__()
+        if namespace:                       # override the default
+            self.namespace = namespace
+        self.events = OrderedDict()
+        self.lock = threading.Lock()
+        self.active = None
+        self.resource_version = None
+
+    def fetch(self):
+        # clear the cache on every call to fetch
+        self.events.clear()
+        try:
+            resp = self.api.list_namespaced_event(self.namespace)
+        # TODO - test for auth problem to be more specific in the except clause
+        except:
+            self.active = False
+            self.health = "[WRN] Unable to access events API (list_namespaced_event call failed)"
+            log.warning(self.health)
+        else:
+            self.active = True
+            self.resource_version = resp.metadata.resource_version
+
+            # TODO this fetches the events, but storing them like this is not preserving the 
+            # creation timestamp of the event itself - it's just using arrival order
```

Comment:
How will the arrival order be preserved? Is it not just sorted by name of the event?

Reply:
I'm not sure this is an issue. I use the event name as the key, and when the events are shown (ceph k8sevents list) the events are first sequenced by last_timestamp (most recent first), so I don't need to worry about storing in sequence. 

If this makes sense to you, I'll remove the TODO comment!

Reply:
Yep, it's just the TODO comment that needs updating then

Reply:
ACK